ddtrace/settings.py

```diff
@@ -46,14 +45,116 @@ def get_from(self, obj):
 
         return pin._config
 
-    def _add(self, integration, settings):
+    def _add(self, integration, settings, merge=True):
         """Internal API that registers an integration with given default
         settings.
 
         :param str integration: The integration name (i.e. `requests`)
         :param dict settings: A dictionary that contains integration settings;
             to preserve immutability of these values, the dictionary is copied
             since it contains integration defaults.
+        :param bool merge: Whether to merge any existing settings with those provided,
+            or if we should overwrite the settings with those provided;
+            Note: when merging existing settings take precedence.
+        """
+        # DEV: Use `getattr()` to call our `__getattr__` helper
+        existing = getattr(self, integration)
+        settings = deepcopy(settings)
+
+        if merge:
+            # DEV: This may appear backwards keeping `existing` as the "source" and `settings` as
+            #   the "destination", but we do not want to let `_add(..., merge=True)` overwrite any
+            #   existing settings
+            #
+            # >>> config.requests['split_by_domain'] = True
+            # >>> config._add('requests', dict(split_by_domain=False))
+            # >>> config.requests['split_by_domain']
+            # True
+            self._config[integration] = IntegrationConfig(self, deepmerge(existing, settings))
+        else:
+            self._config[integration] = IntegrationConfig(self, settings)
+
+    def trace_headers(self, whitelist, blacklist=None):
+        self._http.trace_headers(whitelist, blacklist=blacklist)
+        return self
+
+    def header_is_traced(self, header_name):
+        return self._http.header_is_traced(header_name)
+
+    def __repr__(self):
+        cls = self.__class__
+        integrations = ', '.join(self._config.keys())
+        return '{}.{}({})'.format(cls.__module__, cls.__name__, integrations)
+
+
+class IntegrationConfig(dict):
```

Comment:
This is fine, but we should probably have a bigger config discussion before adding additional classes here.

No reason to remove, we can use this as a good reference for that conversation.

Reply:
Totally agree, let's see what others' guts say